### DISCUSSION TITLE
Scanning devices and getting data together all the time

### DISCUSSION THREAD

**Discussion by FawadRahim** [hbldh/bleak#1008] [Q&A]
I am a newbie to bleak and asynchronous programming, so i would like some suggestions or help i am stuck in one research task where i want that bleak module should scan for devices for 1 minute and give a list in that list the devices should connect and receive data together but if some devices goes disconnected after being turned off then the other devices with recieving notifications/data should not be disturbed. and if the devices again came turned on so it should join other devices to receive data as well. Meantime the scanning should be done after one minute and this will be forever scanning after one minute.

i took some help from the issue section and used a code from a person where i am able to connect to 8 devices at same time and receiving notifications and also i implement the scanning part as well the scanning is happening after 1minute but the problem is once i tuned off one device and turned back on the program dont recognize (this part of code remains).

I need help please from seniors i am very new in field currently i am also studying asyncio programming to understand everything.
here is my code which so far scans and connect to multiple devices at same time while scanning. but for the sake of explaining my code i defined one target device which i want to turn on during scanning to receive notifications and want to turn off and then again wants to turn on so it should not disturb scanning and other devices to receive data. (right now other devices addresses are not mentioned)

```python
import asyncio
from bleak import BleakClient
from functools import partial
import bitstruct
from bleak import BleakScanner, BleakError
from bleak import BleakClient
import datetime
from operator import itemgetter
import logging
import time

NETWORK_NODE_SERVICE_UUID = '680c21d9-c946-4c1f-9c11-baa1c21329e7'  # The format here does not do anything
`#UUID`
OPERATION_MODE_CHARACTERISTIC_UUID = '3f0afd88-7770-46b0-b5e7-9fc099598964'
NETWORK_ID_CHARACTERISTIC_UUID = '80f9d8bc-3bff-45bb-a181-2d6a37991208'
LOCATION_DATA_MODE_CHARACTERISTIC_UUID = 'a02b947e-df97-4516-996a-1882521e0ead'
LOCATION_DATA_CHARACTERISTIC_UUID = '003bbdf2-c634-4b3d-ab56-7ec889b89a37'
PROXY_POSITIONS_CHARACTERISTIC_UUID = 'f4a67d7d-379d-4183-9c03-4b6ea5103291'
DEVICE_INFO_CHARACTERISTIC_UUID = '1e63b1eb-d4ed-444e-af54-c1e965192501'
STATISTICS_CHARACTERISTIC_UUID = '0eb2bc59-baf1-4c1c-8535-8a0204c69de5'
FW_UPDATE_PUSH_CHARACTERISTIC_UUID = '5955aa10-e085-4030-8aa6-bdfac89ac32b'
FW_UPDATE_POLL_CHARACTERISIC_UUID = '9eed0e27-09c0-4d1c-bd92-7c441daba850'
DISCONNECT_CHARACTERISTIC_UUID = 'ed83b848-da03-4a0a-a2dc-8b401080e473'
ANCHOR_PERSISTED_POSITION_CHARACTERISTIC_UUID = 'f0f26c9b-2c8c-49ac-ab60-fe03def1b40c'
ANCHOR_CLUSTER_INFO_CHARACTERISTIC_UUID = '17b1613e-98f2-4436-bcde-23af17a10c72'
ANCHOR_MAC_STATS_CHARACTERISTIC_UUID = '28d01d60-89de-4bfa-b6e9-651ba596232c'
ANCHOR_LIST_CHARACTERISTIC_UUID = '5b10c428-af2f-486f-aee1-9dbd79b6bccb'
TAG_UPDATE_RATE_CHARACTERISTIC_UUID = '7bd47f30-5602-4389-b069-8305731308b6'

ble_devices = []
targetdevice= "C0:E9:BE:C4:08:52" #this is the device which i want to turn on and off during scanning

async def scannerforever():
    global ble_devices
    while True:
        async with BleakScanner() as scanner:
            await asyncio.sleep(60.0)
        for d in scanner.discovered_devices:
            print(d)
            ble_devices.append(d.address)
            print(ble_devices)

def run(addresses):
    
    loop = asyncio.get_event_loop()
    loop1 = asyncio.get_event_loop()
    
    #mycode
    loop2 = asyncio.get_event_loop()

    for address in ble_devices:
        try:
            loop2.create_task(scannerforever())
            if targetdevice == address: #matching the address from scanned ble_devices to check whether the device is turned on or not
                ble_devices.clear() #clearing the list of the ble_devices so that when new scanning performs list should be clear
                loop.create_task(connect_to_device(targetdevice))
            else:
                ble_devices.clear()



        except Exception as e:
            
            loop.stop()
            address = address
            
            loop1.create_task(connect_to_device(address))

    loop.run_forever()
    loop1.run_forever()
    loop2.run_forever()

async def connect_to_device(address):
    async with BleakClient(address, timeout=20.0, use_cached=False) as client:
        print(f"Connected: {address} {client.is_connected}")
        await client.start_notify(LOCATION_DATA_CHARACTERISTIC_UUID, partial(callback, client)) #the callback part is ok i am receiving different clients so that i can know which data is coming from which device
        
        while True:
            await asyncio.sleep(1.0) #to receive data forever

if __name__ == "__main__":
    run(
        [
            targetdevice,
        ]
    )
```

**Comment by dlech** [comment]
The program should have a single top-level `asynio.run(main())` instead of running multiple loops. Do not stop the loop unless you are exiting the program.

And as noted in the troubleshooting section of the docs, scanning and connecting at the same time doesn't tend to work well. It will probably work better if you scan until you find one device, then stop scanning, then connect to that device, then start scanning again.

There are more efficient ways to wait forever, like `await asyncio.Event().wait()`.
